### PR TITLE
Adds an API for landmarks.

### PR DESCRIPTION
**Pull request by pifon2a**

### CODE REVIEW

**Comment by wohe** [comment]
General comment: Please split this into smaller PRs, e.g. start with one that just contains the data structure and the proto message. This leads to quicker reviews and merges.

File: cartographer/sensor/proto/sensor.proto [hunk 1]
```diff
@@ -44,3 +44,10 @@ message FixedFramePoseData {
   optional int64 timestamp = 1;
   optional transform.proto.Rigid3d pose = 2;
 }
+
+// Proto representation of ::cartographer::sensor::LandmarkData.
+message LandmarkData {
+  optional string id = 1;
```
Comment (wohe) [inline cartographer/sensor/proto/sensor.proto]: Should this not include a `timestamp` similar to `FixedFramePoseData`?

We have to be able to report the detection of multiple or no landmarks, e.g. if we detect no landmarks we still have to provide sensor data to allow the SLAM algorithm to advance. Use a `repeated` field of observation message?

File: cartographer/mapping/global_trajectory_builder_interface.h [hunk 2]
```diff
@@ -54,6 +55,8 @@ class GlobalTrajectoryBuilderInterface {
   virtual void AddRangefinderData(common::Time time,
                                   const Eigen::Vector3f& origin,
                                   const sensor::PointCloud& ranges) = 0;
+  virtual void AddLandmarkData(common::Time time,
```
Comment (wohe) [inline cartographer/mapping/global_trajectory_builder_interface.h]: It's better to follow the pattern using `AddSensorData` which allows you to use the `Dispatchable<>` template.

**Comment by jihoonl** [comment]
Just from curiosity. It doesn't seem to consider the orientation of landmark. Is it intended?

**Comment by pifon2a** [comment]
It will, adding them right now.

**Comment by wohe** [comment]
This contains a lot of unrelated changes that make it hard to review. Rebase?

**Comment by pifon2a** [comment]
[](https://i.ytimg.com/vi/fK1_hjviEl8/maxresdefault.jpg)

Sorry, I am still not used to git. Done.

**Comment by timethy** [comment]
Hi, I would be happy to help contributing on this PR.
@pifon2a please contact me via mail on taubnert@student.ethz.ch if you're up for some discussion on how I could be of any help.

File: cartographer/sensor/proto/sensor.proto [hunk 3]
```diff
@@ -44,3 +44,15 @@ message FixedFramePoseData {
   optional int64 timestamp = 1;
   optional transform.proto.Rigid3d pose = 2;
 }
+
+// Proto representation of ::cartographer::sensor::LandmarkData.
+message LandmarkData {
+  message Landmark {
+    optional string id = 1;
+    optional transform.proto.Vector3d translation = 2;
```
Comment (wohe) [inline cartographer/sensor/proto/sensor.proto]: Use `transform.proto.Rigid3d`?

File: cartographer/sensor/proto/sensor.proto [hunk 4]
```diff
@@ -44,3 +44,15 @@ message FixedFramePoseData {
...
+    optional transform.proto.Quaterniond alignment = 3;
+    optional double weight = 4;
```
Comment (wohe) [inline cartographer/sensor/proto/sensor.proto]: We should have separate `translation_weight` and `rotation_weight`.

**Pull request closed without merging**